pkg/pillar/pubsub/pubsub.go

```diff
@@ -6,455 +6,68 @@
+//  s1 := pubsub.Subscribe("foo", fooStruct{}, false, &myctx)
+//  s1.ModifyHandler = func(...), // Optional
+//  s1.DeleteHandler = func(...), // Optional
+//  s1.RestartHandler = func(...), // Optional
+//  [ Initialize myctx ]
+//  s1.Activate()
+//  ...
+//  select {
+//     case change := <- s1.C:
+//         s1.ProcessChange(change, ctx)
+//  }
+type SubHandler func(ctx interface{}, key string, status interface{})
+type SubRestartHandler func(ctx interface{}, restarted bool)
+
+type Subscription interface {
+	Channel() <-chan string
+	Activate() error
+	ProcessChange(change string)
+	Get(key string) (interface{}, error)
+	GetAll() map[string]interface{}
+	Restarted() bool
+	Synchronized() bool
+	Handlers() (SubHandler, SubHandler, SubHandler, SubRestartHandler, SubRestartHandler)
+	SetHandlers(create, modify, delete SubHandler, restart, synchronized SubRestartHandler)
 }
 
 // Init function for Subscribe; returns a context.
 // Assumption is that agent with call Get(key) later or specify
 // handleModify and/or handleDelete functions
 // watch ensures that any restart/restarted notification is after any other
 // notifications from ReadDir
-func Subscribe(agentName string, topicType interface{}, activate bool,
-	ctx interface{}) (*Subscription, error) {
-
-	return subscribeImpl(agentName, "", topicType, activate, ctx, false)
-}
-
-func SubscribeScope(agentName string, agentScope string, topicType interface{},
-	activate bool, ctx interface{}) (*Subscription, error) {
-
-	return subscribeImpl(agentName, agentScope, topicType, activate, ctx,
-		false)
-}
-
-func SubscribePersistent(agentName string, topicType interface{}, activate bool,
-	ctx interface{}) (*Subscription, error) {
-
-	return subscribeImpl(agentName, "", topicType, activate, ctx, true)
-}
-
-func subscribeImpl(agentName string, agentScope string, topicType interface{},
-	activate bool, ctx interface{}, persistent bool) (*Subscription, error) {
-
-	topic := TypeToName(topicType)
-	changes := make(chan string)
-	sub := new(Subscription)
-	sub.C = changes
-	sub.sendChan = changes
-	sub.topicType = topicType
-	sub.agentName = agentName
-	sub.agentScope = agentScope
-	sub.topic = topic
-	sub.userCtx = ctx
-	sub.km = keyMap{key: NewLockedStringMap()}
-	sub.persistent = persistent
-	name := sub.nameString()
 
-	// Special case for files in /var/tmp/zededa/ and also
-	// for zedclient going away yet metrics being read after it
-	// is gone.
-	if agentName == "" {
-		sub.subscribeFromDir = true
-		sub.dirName = FixedDirName(name)
-	} else if agentName == "zedclient" {
-		sub.subscribeFromDir = true
-		sub.dirName = PubDirName(name)
-	} else if persistent {
-		sub.subscribeFromDir = true
-		sub.dirName = PersistentDirName(name)
-	} else {
-		sub.subscribeFromDir = subscribeFromDir
-		sub.dirName = PubDirName(name)
-	}
-	log.Infof("Subscribe(%s)\n", name)
-	if activate {
-		if err := sub.Activate(); err != nil {
-			return nil, err
-		}
-	}
-	return sub, nil
+type PubSub interface {
```

Comment:
Why do we need this? Can't we just do with Publish and Subscribe?

Reply:
That is a bigger discussion. I will answer in the regular comments. I think you may have a point here.